sdk/nodejs/resource.ts

```diff
@@ -676,3 +676,160 @@ export class ComponentResource extends Resource {
 export const testingOptions = {
     isDryRun: false,
 };
+
+
+/**
+ * [mergeOptions] takes two ResourceOptions values and produces a new ResourceOptions with the
+ * respective properties of `opts2` merged over the same properties in `opts1`.  The original
+ * options objects will be unchanged.
+ *
+ * Conceptually property merging follows these basic rules:
+ *  1. if the property is a collection, the final value will be a collection containing the values
+ *     from each options object.
+ *  2. Simple scaler values from `opts2` (i.e. strings, numbers, bools) will replace the values of
+ *     `opts1`.
+ *  3. `opts2` can have properties explicitly provided with `null` or `undefined` as the value. If
+ *     explicitly provided, then that will be the final value in the result.
+ *  4. For the purposes of merging `dependsOn`, `provider` and `providers` are always treated as
+ *     collections, even if only a single value was provided.
+ */
+export function mergeOptions(opts1: CustomResourceOptions | undefined, opts2: CustomResourceOptions | undefined): CustomResourceOptions;
+export function mergeOptions(opts1: ComponentResourceOptions | undefined, opts2: ComponentResourceOptions | undefined): ComponentResourceOptions;
+export function mergeOptions(opts1: ResourceOptions | undefined, opts2: ResourceOptions | undefined): ResourceOptions;
+export function mergeOptions(opts1: ResourceOptions | undefined, opts2: ResourceOptions | undefined): ResourceOptions {
+    const dest = <any>{ ...opts1 };
+    const source = <any>{ ...opts2 };
+
+    // Ensure provider/providers are all expanded into the `{ provName: prov }` form.
+    // This makes merging simple.
+    expandProviders(dest);
+    expandProviders(source);
+
+    // iterate specifically over the supplied properties in [source].  Note: there may not be an
+    // corresponding value in [dest].
+    for (const key of Object.keys(source)) {
+        const destVal = dest[key];
+        const sourceVal = source[key];
+
+        if (key === "providers") {
+            // Note: this expansion is safe to do as expandProviders will have made sure
+            // that both collections are in Record<string, ProviderResource> form.
+            dest[key] = { ...destVal, ...sourceVal };
+            continue;
+        }
+
+        // Due to the possibility of top level and nested Inputs for 'dependsOn' we have to handle
+        // that property specially.
+        if (key === "dependsOn") {
+            dest[key] = mergeDependsOn(destVal, sourceVal);
+            continue;
+        }
+
+        // we should have no promises/outputs at top level for any other resource option properties.
+        if (isPromiseOrOutput(destVal)) {
+            throw new Error(`Unexpected promise/output in opts1.${key}`);
```

Comment:
What happens when we add new properties in the future?  Especially in cases of sxs where the version of this function used is older than the version of reaourceoptions?

Feels like we may want to be more liberal and/or generic here?

Reply:
> Feels like we may want to be more liberal and/or generic here?

The problem is that i literally don't know how to make htis more generic.  The rules are not simple.  For example, we do *not* unwrap the underlying values in "aliases".  There, we keep things as arrays (so that we can merge parent/child arrays as we hit them), and thus don't want an array containing a promise to become a promise at the top leve.

Reply:
let me see waht i can do here to make this better.

Reply:
I think i've been able to simplify this a bunch.  Please take a look when you can.